File: javascript/data-structures/p236_Deque.js
/*
    CLRS Exercise 10.1-5, p. 236
*/

class Deque {
    constructor(Q = [], head = 0, tail = 0) {
        this.Q = Q;
        this.head = head;
        this.tail = tail;
    }

    dequeEmpty = () => this.head === this.tail;

    dequeFull = () => (this.head === this.tail + 1) 
                        || (this.head === 0 && this.tail === this.Q.length - 1);

    headEnqueue = (x) => {
        if (this.dequeFull()) {
            throw new Error('overflow');
        } else {
            if (this.head === 0) {
                this.head = this.Q.length - 1;
            } else {
                this.head--;
            }
            this.Q[this.head] = x;
        }
    }

    tailEnqueue = (x) => {
        if (this.dequeFull()) {
            throw new Error('overflow');
        } else {
            this.Q[this.tail] = x;
            if (this.tail === this.Q.length - 1) {
                this.tail = 0;
            } else {
                this.tail++;
            }
        }
    }

    headDequeue = () => {
        if(this.dequeEmpty()) {
            throw new Error('underflow');
        } else {
            const x = this.Q.splice(this.head, 1, undefined)[0];
            if (this.head === this.Q.length - 1) {
                this.head = 0;
            } else {
                this.head++;
            }
            return x;
        }
    }

    tailDequeue = () => {
        if(this.dequeEmpty()) {
            throw new Error('underflow');
        } else {
            if (this.tail === 0) {
                this.tail = this.Q.length - 1;
            } else {
                this.tail--;
            }
            return this.Q.splice(this.tail, 1, undefined)[0];
        }
    }
}

module.exports = {
    Deque
}

File: python/data_structures/p236_deque.py
'''
    CLRS Exercise 10.1-5, p. 236
'''

class Deque:
    def __init__(self, Q=None, head=0, tail=0):
        if Q is None:
            Q = []
        else:
            self.Q = Q
        self.head = head
        self.tail = tail

    def deque_empty(self):
        return self.head == self.tail
    
    def deque_full(self):
        return (self.head == self.tail + 1) \
            or (self.head == 0 and self.tail == len(self.Q) - 1)
    
    def head_enqueue(self, x):
        if self.deque_full():
            raise Exception('overflow')
        else:
            if self.head == 0:
                self.head = len(self.Q) - 1
            else:
                self.head -= 1
            self.Q[self.head] = x
    
    def tail_enqueue(self, x):
        if self.deque_full():
            raise Exception('overflow')
        else:
            self.Q[self.tail] = x
            if self.tail == len(self.Q) - 1:
                self.tail = 0
            else:
                self.tail += 1
    
    def head_dequeue(self):
        if self.deque_empty():
            raise Exception('underflow')
        else:
            x = self.Q[self.head]; self.Q[self.head] = None
            if self.head == len(self.Q) - 1:
                self.head = 0
            else:
                self.head += 1
            return x
    
    def tail_dequeue(self):
        if self.deque_empty():
            raise Exception('underflow')
        else:
            if self.tail == 0:
                self.tail = len(self.Q) - 1
            else:
                self.tail -= 1
                self.Q.insert(self.tail, None)
                return self.Q.pop(self.tail + 1)
